pkg/reconciler/ingress/ingress.go

```diff
@@ -0,0 +1,279 @@
+	networkingv1 "k8s.io/api/networking/v1"
+	"k8s.io/apimachinery/pkg/api/errors"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/labels"
+	"k8s.io/client-go/tools/cache"
+	"k8s.io/klog/v2"
+)
+
+const (
+	clusterLabel = "kcp.dev/cluster"
+	ownedByLabel = "kcp.dev/owned-by"
+)
+
+// reconcile is triggered on every change to an ingress resource, or it's associated services (by tracker).
+func (c *Controller) reconcile(ctx context.Context, ingress *networkingv1.Ingress) error {
+	klog.Infof("reconciling Ingress %q", ingress.Name)
+
+	if ingress.Labels == nil || ingress.Labels[clusterLabel] == "" {
+		// This is a root Ingress; get its leafs.
+		sel, err := labels.Parse(fmt.Sprintf("%s=%s", ownedByLabel, ingress.Name))
+		if err != nil {
+			return err
+		}
+
+		// Get the current Leaves
+		currentLeaves, err := c.lister.List(sel)
+		if err != nil {
+			return err
+		}
+
+		// Generate the desired leaves
+		desiredLeaves, err := c.desiredLeaves(ctx, ingress)
+		if err != nil {
+			return err
+		}
+
+		// Clean the leaves that are not desired anymore
+		for _, leaftoremove := range findNonDesiredLeaves(currentLeaves, desiredLeaves) {
+			klog.Infof("Deleting non desired leaf %q", leaftoremove.Name)
+			if err := c.client.NetworkingV1().Ingresses(leaftoremove.Namespace).Delete(ctx, leaftoremove.Name, metav1.DeleteOptions{}); err != nil {
+				return err
+			}
+		}
+
+		// TODO(jmprusi): ugly. fix. use indexer, etc.
+		// Create and/or update the desired leaves
+		for _, desiredleaf := range desiredLeaves {
+			if _, err := c.client.NetworkingV1().Ingresses(desiredleaf.Namespace).Create(ctx, desiredleaf, metav1.CreateOptions{}); err != nil {
+				if errors.IsAlreadyExists(err) {
+					existingLeaf, err := c.client.NetworkingV1().Ingresses(desiredleaf.Namespace).Get(ctx, desiredleaf.Name, metav1.GetOptions{})
+					if err != nil {
+						return err
+					}
+
+					// Set the resourceVersion and UID to update the desired leaf.
+					desiredleaf.ResourceVersion = existingLeaf.ResourceVersion
+					desiredleaf.UID = existingLeaf.UID
+
+					if _, err := c.client.NetworkingV1().Ingresses(desiredleaf.Namespace).Update(ctx, desiredleaf, metav1.UpdateOptions{}); err != nil {
+						return err
+					}
+
+				} else {
+					return err
+				}
+			}
+		}
+
+	} else {
+		// If the ingress has the clusterLabel set, that means that it is a leaf and it's synced with
+		// a cluster.
+		//
+		// This update can come from the creation or because the syncer has update the status.
+
+		rootIngressName := ingress.Labels[ownedByLabel]
+		// A leaf Ingress was updated; get others and aggregate status.
+		sel, err := labels.Parse(fmt.Sprintf("%s=%s", ownedByLabel, rootIngressName))
+		if err != nil {
+			return err
+		}
+		others, err := c.lister.List(sel)
+		if err != nil {
+			return err
+		}
+
+		// Get the rootIngress based on the labels.
+		var rootIngress *networkingv1.Ingress
+
+		rootIf, exists, err := c.indexer.Get(&networkingv1.Ingress{
```

Comment:
If you make the change so the value of the ownedByLabel is the full key, you can do c.indexer.GetByKey(rootIngressName) here.